pallets/rewards/src/mechanism/gap.rs

```diff
@@ -0,0 +1,377 @@
+
+		type IBalance: FixedPointOperand
+			+ TryFrom<Self::Balance>
+			+ codec::FullCodec
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ EnsureAdd
+			+ EnsureSub
+			+ Copy
+			+ Signed
+			+ sp_std::fmt::Debug
+			+ Default;
+
+		type Rate: FixedPointNumber + codec::FullCodec + TypeInfo + MaxEncodedLen;
+
+		type MaxCurrencyMovements: Get<u32> + sp_std::fmt::Debug;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type RptHistory<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::DistributionId, T::Rate>;
+
+	#[pallet::storage]
+	pub(super) type LastDistributionId<T: Config> = StorageValue<_, T::DistributionId, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		// Emits when a currency is moved but any account associated has pending stake.
+		// Currency movement is only allowed after a distribution, with no deposit/withdraw stake
+		// from any participant.
+		TryMovementAfterPendingState,
+	}
+
+	impl<T: Config> RewardMechanism for Pallet<T>
+	where
+		<T::Rate as FixedPointNumber>::Inner: Signed,
+	{
+		type Account = Account<T>;
+		type Balance = T::Balance;
+		type Currency = Currency<T>;
+		type Group = Group<T>;
+		type MaxCurrencyMovements = T::MaxCurrencyMovements;
+
+		fn reward_group(
+			group: &mut Self::Group,
+			amount: Self::Balance,
+		) -> Result<Self::Balance, DispatchError> {
+			let mut reward = Self::Balance::zero();
+
+			if group.total_stake > T::Balance::zero() {
+				let rate = T::Rate::ensure_from_rational(amount, group.total_stake)?;
+				group.rpt.ensure_add_assign(rate)?;
+				reward = amount;
+			}
+
+			group
+				.total_stake
+				.ensure_add_assign(group.pending_total_stake)?;
+			group.pending_total_stake = T::Balance::zero();
+
+			RptHistory::<T>::insert(group.distribution_id, group.rpt);
+
+			group.distribution_id = LastDistributionId::<T>::try_mutate(
+				|distribution_id| -> Result<T::DistributionId, DispatchError> {
+					Ok(*distribution_id.ensure_add_assign(One::one())?)
+				},
+			)?;
+
+			Ok(reward)
+		}
+
+		fn deposit_stake(
+			account: &mut Self::Account,
+			currency: &mut Self::Currency,
+			group: &mut Self::Group,
+			amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			account.update(group, currency)?;
+
+			account.pending_stake.ensure_add_assign(amount)?;
+			group.pending_total_stake.ensure_add_assign(amount)?;
+
+			currency.total_stake.ensure_add_assign(amount)?;
+
+			Ok(())
+		}
+
+		fn withdraw_stake(
+			account: &mut Self::Account,
+			currency: &mut Self::Currency,
+			group: &mut Self::Group,
+			amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			account.update(group, currency)?;
+
+			let pending_amount = amount.min(account.pending_stake);
+
+			account.pending_stake.ensure_sub_assign(pending_amount)?;
+			group
+				.pending_total_stake
+				.ensure_sub_assign(pending_amount)?;
+
+			let computed_amount = amount.ensure_sub(pending_amount)?;
```

Comment:
Maybe remaining_amount? Since it is what is left after subtracting from pending.
```suggestion
			let remaining_amount = amount.ensure_sub(pending_amount)?;
```

Reply:
I call it computed, because it is the part of the amount that will be subtracted from the already computed stake for reward of the account. Not sure if "remaining" indicates that, maybe too generic?

If computed is not descriptive enough, maybe something like `rewarded_amount` fits better?